File: sdks/java/src/main/java/com/apollodeploy/signal/sdk/api/MetricsAPI.java
package com.apollodeploy.signal.sdk.api;

import com.fasterxml.jackson.core.type.TypeReference;
import com.apollodeploy.signal.sdk.models.Types.*;
import com.apollodeploy.signal.sdk.transport.Transport;
import java.time.Duration;
import java.util.List;

/** MetricsAPI API operations. */
public final class MetricsAPI {
    private final Transport transport;

    public MetricsAPI(Transport transport) {
        this.transport = transport;
    }

    public TopicPerformanceResponse getTopicPerformance(
        String id,
        String window,
        String projectId
    ) {
        String path = "/v1/metrics/topics/${id}";
        path = path.replace("${id}", Transport.encodePathSegment(String.valueOf(id)));

        Transport.Request request = new Transport.Request("GET", path);
        if (window != null) {
            request.query("window", window);
        }
        if (projectId != null) {
            request.query("projectId", projectId);
        }
        return transport.execute(request, new TypeReference<TopicPerformanceResponse>() {
        });
    }

    public EmailPerformanceResponse getEmailPerformance(
        String id
    ) {
        String path = "/v1/metrics/emails/${id}";
        path = path.replace("${id}", Transport.encodePathSegment(String.valueOf(id)));

        Transport.Request request = new Transport.Request("GET", path);
        return transport.execute(request, new TypeReference<EmailPerformanceResponse>() {
        });
    }

    public EmailReadScrollAnalyticsResponse getEmailEngagement(
        String id
    ) {
        String path = "/v1/metrics/emails/${id}/engagement";
        path = path.replace("${id}", Transport.encodePathSegment(String.valueOf(id)));

        Transport.Request request = new Transport.Request("GET", path);
        return transport.execute(request, new TypeReference<EmailReadScrollAnalyticsResponse>() {
        });
    }

    public ProjectMetricsSummaryResponse getProjectMetricsSummary(
        String window,
        String projectId
    ) {
        String path = "/v1/metrics/summary";

        Transport.Request request = new Transport.Request("GET", path);
        if (window != null) {
            request.query("window", window);
        }
        if (projectId != null) {
            request.query("projectId", projectId);
        }
        return transport.execute(request, new TypeReference<ProjectMetricsSummaryResponse>() {
        });
    }

    public ProjectMetricsTimelineResponse getProjectMetricsTimeline(
        String window,
        String granularity,
        String format,
        String projectId
    ) {
        String path = "/v1/metrics/timeline";

        Transport.Request request = new Transport.Request("GET", path);
        if (window != null) {
            request.query("window", window);
        }
        if (granularity != null) {
            request.query("granularity", granularity);
        }
        if (format != null) {
            request.query("format", format);
        }
        if (projectId != null) {
            request.query("projectId", projectId);
        }
        return transport.execute(request, new TypeReference<ProjectMetricsTimelineResponse>() {
        });
    }

    public AdvisorReportResponse getMetricsAdvisor(
        String projectId
    ) {
        String path = "/v1/metrics/advisor";

        Transport.Request request = new Transport.Request("GET", path);
        if (projectId != null) {
            request.query("projectId", projectId);
        }
        return transport.execute(request, new TypeReference<AdvisorReportResponse>() {
        });
    }

}

File: sdks/kotlin/src/main/kotlin/com/apollodeploy/signal/sdk/api/MetricsAPI.kt
package com.apollodeploy.signal.sdk.api

import com.apollodeploy.signal.sdk.internal.Transport
import com.apollodeploy.signal.sdk.models.*
import io.ktor.http.HttpMethod
import io.ktor.http.encodeURLPathPart
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.coroutines.flow.Flow

/** MetricsAPI API operations. */
class MetricsAPI(
    private val transport: Transport,
) {
/** getTopicPerformance operation. */
    suspend fun getTopicPerformance(
        id: String,
        window: String? = null,
        projectId: String? = null,
    ): TopicPerformanceResponse {
        var path = "/v1/metrics/topics/\${id}"
        path = path.replace("\${id}", id.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        window?.let { query += "window" to it.toString() }
        projectId?.let { query += "projectId" to it.toString() }
        val headers = mutableMapOf<String, String>()

        return transport.execute<TopicPerformanceResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getEmailPerformance operation. */
    suspend fun getEmailPerformance(
        id: String,
    ): EmailPerformanceResponse {
        var path = "/v1/metrics/emails/\${id}"
        path = path.replace("\${id}", id.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<EmailPerformanceResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getEmailEngagement operation. */
    suspend fun getEmailEngagement(
        id: String,
    ): EmailReadScrollAnalyticsResponse {
        var path = "/v1/metrics/emails/\${id}/engagement"
        path = path.replace("\${id}", id.toString().encodeURLPathPart())

        val query = mutableListOf<Pair<String, String>>()
        val headers = mutableMapOf<String, String>()

        return transport.execute<EmailReadScrollAnalyticsResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getProjectMetricsSummary operation. */
    suspend fun getProjectMetricsSummary(
        window: String? = null,
        projectId: String? = null,
    ): ProjectMetricsSummaryResponse {
        var path = "/v1/metrics/summary"

        val query = mutableListOf<Pair<String, String>>()
        window?.let { query += "window" to it.toString() }
        projectId?.let { query += "projectId" to it.toString() }
        val headers = mutableMapOf<String, String>()

        return transport.execute<ProjectMetricsSummaryResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getProjectMetricsTimeline operation. */
    suspend fun getProjectMetricsTimeline(
        window: String? = null,
        granularity: String? = null,
        format: String? = null,
        projectId: String? = null,
    ): ProjectMetricsTimelineResponse {
        var path = "/v1/metrics/timeline"

        val query = mutableListOf<Pair<String, String>>()
        window?.let { query += "window" to it.toString() }
        granularity?.let { query += "granularity" to it.toString() }
        format?.let { query += "format" to it.toString() }
        projectId?.let { query += "projectId" to it.toString() }
        val headers = mutableMapOf<String, String>()

        return transport.execute<ProjectMetricsTimelineResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

/** getMetricsAdvisor operation. */
    suspend fun getMetricsAdvisor(
        projectId: String? = null,
    ): AdvisorReportResponse {
        var path = "/v1/metrics/advisor"

        val query = mutableListOf<Pair<String, String>>()
        projectId?.let { query += "projectId" to it.toString() }
        val headers = mutableMapOf<String, String>()

        return transport.execute<AdvisorReportResponse>(
            method = HttpMethod.Get,
            path = path,
            query = query.ifEmpty { null },
            body = null,
            headers = headers.ifEmpty { null },
            operationTimeoutMs = null,
        )
    }

}
